mautic/mautic
# Manage length of custom field values

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | No
| New feature/enhancement? (use the a.x branch)      | Yes
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | No
| Automated tests included?              | Yes 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:
We're pretty consistently hitting custom field limits with our larger customers. It's because of a MySQL hard limit that we can't ultimately avoid. However, we can help mitigate it by adding a couple changes to Mautic. Currently we create all "text" fields with a 255 characters length and attempt to add an index to each.

We should implement the feature to add the ability to define the maximum number of characters for a "text" custom field. That way the customer can decide if the field only has values with a max of 10 characters which would then give 245 characters to another field.

We provide a "indexable" flag to the "text" custom field. If set to yes, they must define the limit as less than 255 characters and we can also add some validation to determine if we will hit MySQL limits for indexes (only 64 are allowed). If they have a field that is more than 255 characters and/or not indexable, then we can create the field as a MySQL text field instead of a varchar which is not supposed to go against the hard row length limit. We should also create textarea custom fields as MySQL text fields so they are not counted toward the length.

Decide what should be "Indexable" translated to
Show index flag on custom field list in last column as icon
Filter by indexed flag
Add indexable flag to text column field
use it when creating/editing custom field
Show in popup what "indexable means with available indexes count
Validate count of indexes when creating/editing custom fields
limit character length less than 255 via additional input if indexable
use it when updating data
Sync saved indexed w existing indexes
Change multi select to long text (no migration)
Don't allow editing length (but allow editing indexes)
Add migration to set text fields to 255 length (BC)


#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Run migrations, new columns is_index and char_length_limit in lead_fields table are created and char_length_limit is set to 255 for current text fields, is_index is set to true for current fields if the field has already index
3. When creating text fields (text, select, phone, url, email), you can use indexable flag to decide whether the index should be created and you can set field length, which will be used for column creation in the db. Default value can't be longer than character limit. Indexable flag has a tooltip with information about remaining available indexes.
4. You can change indexable flag for current fields, index will be updated in the db
5. While submitting contact, text fields are validated for a character limit
6. Filters are fixed for custom fields, you can filter also indexable or unique fields

**Add/Update custom field with both value true and false for create_custom_field_in_background flag**
